### ISSUE TITLE
Semitransparent clouds have ugly inner corners

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#5858]
If you use the clouds to make semi-transparent clouds with a low opacity, the inner corners of clouds become more and more ugly:

![Showing cloud with bad inner corners](https://i.imgur.com/4siopcx.png)
![Another example](https://i.imgur.com/Bj8vevr.png)

Tested in: https://github.com/minetest/minetest/commit/1681a009bc54b19eeab0356c7ed856bc0bed6a1a

**Comment by paramat** [comment]
Not a bug, it's just translucency adding at points where you are looking through 2 masses of cloud. I like it, it creates a more 3D effect. Although yes the addition seems exaggerated at low opacity, but this is likely an Irrlicht behaviour.

**Comment by Wuzzy2** [comment]
> Interesting observation: If you move very close to one of these inner corners, the visible glitch disappears and it looks like it should.

Now if the overlap is intended, than *that* above is a bug.

**Comment by paramat** [comment]
> Interesting observation: If you move very close to one of these inner corners, the visible glitch disappears 

Yes i agree that behaviour is a bug.

The overlap opacity does seem a little too much when using a low opacity, so maybe there is some way to reduce the strength, i would not want to remove the effect completely though.

**Comment by paramat** [comment]
Maybe a translucency issue due to MT's translucency problems?
@bendeutsch any comments?

**Comment by bendeutsch** [comment]
I have no clue, sorry. The original behavior, the overlap, is exactly what I'd expect from a simple OpenGL-like drawing of simple polygons. My knowledge of such systems is that non-transparent polygons are rendered first, using a z-buffer to handle occlusions. Then the transparent polygons are rendered, without a z-buffer (since you actually have to draw all of them) but sorted back-to-front, the "painter's algorithm".

To get an effect like the close-ups, my best guess is that Irrlicht renders the polygons front-to-back, _with_ a z-buffer, which prevents the overlapped part of the inner corner from being drawn at all. The engine would need to know that these polygons "belong to the same thing" and should not be visible through each other. Unfortunately, I don't know Irrlicht nearly enough to say if it even does this, or how the cloud rendering would transmit this to Irrlicht, sorry. Maybe an Irrlicht pro can weigh in?

**Comment by paramat** [comment]
I get the feeling the opacity of overlapping cloud is actually mathematically 'correct' but just looks a little too strong for human visual taste. Not sure we can do anything about it anyway as it might be just how translucencies add up in Irrlicht.

**Comment by Wuzzy2** [comment]
@paramat: Sorry, you don't seem to get the problem. I have no preference for either overlap or no overlap.

The bug is that the cloud corners “randomly” switch from one appearance to another **while you move**. This is definitely wrong, no matter if you consider overlap as such to be correct or incorrect. No matter what rendering you want, at least it should be consistent.

Look at my previous post again.

**Comment by paramat** [comment]
Yeah i know, that's what i meant by 'The disappearing overlap is more of an issue.' :)
